tests/integration_tests/reaper.py

```diff
@@ -0,0 +1,198 @@
+            self.wake_reaper.notify()
+            LOG.info("Reaper: awakened to reap")
+        if self.reaper_thread:
+            self.reaper_thread.join()
+        LOG.info("Reaper: stopped")
+
+    def _destroy(self, instance: IntegrationInstance) -> bool:
+        """destroy() destroys an instance and returns True on success."""
+        try:
+            LOG.info("Reaper: destroying %s", instance.instance.id)
+            instance.destroy()
+            self.counter += 1
+            return True
+        except Exception as e:
+            LOG.warning(
+                "Error while tearing down instance %s: %s ", instance, e
+            )
+            return False
+
+    def _reaper_loop(self) -> None:
+        """reaper_loop() manages all instances that have been reaped
+
+        tasks:
+        - destroy newly reaped instances
+        - manage a ledger undead instances
+        - periodically attempt to kill undead instances
+        - die when instructed to
+        - ensure that every reaped instance is destroyed at least once before
+          reaper dies
+        """
+        LOG.info("Reaper: exalted in life, to assist others in death")
+        while True:
+            # nap until woken or timeout
+            with self.wake_reaper:
+                self.wake_reaper.wait(timeout=self.timeout)
+            if self._do_reap():
+                break
+        LOG.info("Reaper: exited")
+
+    def _do_reap(self) -> bool:
+        """_do_reap does a single pass of the reaper loop
+
+        return True if the loop should exit
+        """
+
+        new_undead_instances: List[IntegrationInstance] = []
+
+        # first destroy all newly reaped instances
+        while not self.reaped_instances.empty():
+            instance = self.reaped_instances.get_nowait()
+            success = self._destroy(instance)
+            if not success:
+                LOG.warning(
+                    "Reaper: failed to destroy %s",
+                    instance.instance.id,
+                )
+                # failure to delete, add to the ledger
+                new_undead_instances.append(instance)
+            else:
+                LOG.info("Reaper: destroyed %s", instance.instance.id)
+
+        # every instance has tried at least once and the reaper has been
+        # instructed to tear down - so do it
+        if self.exit_reaper.is_set():
+            if not self.reaped_instances.empty():
+                # race: an instance was added to the queue after iteration
+                # completed. Destroy the latest instance.
+                self._update_undead_ledger(new_undead_instances)
+                return False
+            self._update_undead_ledger(new_undead_instances)
+            LOG.info("Reaper: exiting")
+            if self.undead_ledger:
+                # undead instances exist - unclean teardown
+                LOG.info(
+                    "Reaper: the faults of incompetent abilities will be "
```

Comment:
This shouldn't fit as well as it does :joy:

Reply:
Heh :)

I was originally planning to remove the weird personification of this background thread, but it grew on me a little. If anyone objects I can remove it.

Reply:
No objections from me!